### ISSUE TITLE
Media files marked as uncompressed cannot be played

### ISSUE THREAD

**Issue by nleme** [readium/readium-sdk#177]
We recently discovered this issue, when we started using a new packager to create EPUB3 books. As we know, all EPUB3 books are ZIP files that contain a number of resources, including media files. Up to this point, media files were "compressed" into the ZIP file, even though the "compression" did not result in any gain, because the vast majority of media formats (like MPEG-4, H.264, etc...) are not only highly compressed, but also take advantage of the specific media to compress it even further.

Because of that, recently we adopted a packager that would mark those resources as "uncompressed", because as we just said, it does not make sense to "compress" media files. And then we discovered the following problem. For uncompressed entries, the libzip code follows a different code path, which currently has a bug when seek is done. That bug is such that the libzip code loses track of the right position in the file. As a result, media files marked as "uncompressed" cannot be played by the Readium SDK, through the Launcher.

**Comment by nleme** [comment]
So, I spent quite some trying to make the latest libzip code work with the Readium SDK. A brief explanation about this: libzip is an Open Source project. More than two years ago, Jim Dovey took a snapshot of the libzip code as it was at the time, and put it into the Readium SDK itself, which started using it to deal with the EPUB3 files (since they are basically ZIP files). Since them, the libzip Open Source project kept evolving, and recently it actually reached version 1.0. The problem, as I find out, is that some of the features that are being used by the Readium SDK are no longer available in libzip. They have been replaced by more sophisticated features, that would require quite some code change in the Readium SDK to make them work. I can testify that it is not easy: I spent quite some time over that and still more changes were needed before I could get anything to work at all.

Unfortunately, I do not have the needed time to make the new version of libzip work. On the other hand, we must fix the uncompressed media playback issue. Fortunately, I had made a fix beforehand in the libzip code in Readium SDK, that takes care of this issue. I made that fix in a separate feature branch, and I created a pull request to have that feature branch integrated into develop.

We can take this fix for now and, later, somebody that has the necessary time can try to update the libzip code in the Readium SDK.

The pull request is at:

https://github.com/readium/readium-sdk/pull/179

**Comment by danielweck** [comment]
I commented in the PR (#179) about @DennisPopovDn 's feature branch (updated zip lib with support for large files). Let's try to figure out what the current status of that work is (what version of the zip lib is used, and whether this fork includes significant or trivial changes compared to the original).

**Comment by nleme** [comment]
I took a look at the changes that Dennis Popov made. He reached the same problem that I did, i.e., that _bytes_left is no longer accessible for us to use as a way to keep track of the read position in a given file. He then tried a different solution: he wrote external code to keep track of the cursor itself. He makes a call to zip_source_stat() when the ZIP file is opened, to find out the total size of the file. Then, at each read operation, he diminishes the value of its own bytes_left, to keep track of the cursor.

I see two issues with that approach. The first is just my fear (which it could be unfounded) that somehow our code would lose sync with the libzip code, so that the cursor that our code is keeping itself would lose track with the actual position kept by libzip code. That is something that we can verify, though.

The biggest issue is the second one. Dennis gave up on trying to make his solution work with Seek. He #ifdef'ed that piece of code out and added the following comment:

#if 0   // this functionality is disabled due to macro incompatibility with the latest libzip

The problem is that we *need* the Seek functionality: the media playback needs it. Therefore, unless we modify Dennis' change to also work with Seek, we cannot take his change.

I also found out that I cannot compile Dennis' change in Xcode. I wonder if he was looking at the issue with libzip solely from a Windows point of view.

The bottom line is that there is still quite work to be done before we can take Dennis' change into the Readium SDK.

**Comment by danielweck** [comment]
Thank you Nelson, this helps a lot.

**Comment by DennisPopovDn** [comment]
@nleme , @danielweck , I've started fixing 'Seek' method support

**Comment by danielweck** [comment]
@DennisPopovDn, this sounds great!

**Comment by nleme** [comment]
@DennisPopovDn - do you have any idea how you are fixing seek?

@danielweck - I found another way around this. I got the latest version of libzip, and I discovered this: libzip actually works perfect with seek and tell if you remove the CRC computation when reading a file. This is not a big deal because in the current code we are not using CRC anyway, therefore we are not worse than before. On the other hand, we can take the latest code for libzip *and* use seek and tell fine. I already write the code for that, and so far it is working great. :-)

I can talk about this more on tomorrow's Readium SDK meeting. It is a *very* complex issue, that probably will work better if I talk to you guys. I'm writing an e-mail to Ric to let him know that tomorrow's (06/24/2015) meeting may be long.

**Comment by DennisPopovDn** [comment]
@nleme, this is great news that you found another way around this!  It seems like this is a kind of combination of the old zip_fseek.c file with the new zip_source_seek.c code? And in this case your approach may be  better than my more complicated one...

My idea was to replace byte_stream:844 zip_fseek(....) call with the new one
```C++
zip_source_seek.c:39
ZIP_EXTERN int zip_source_seek(zip_source_t *src, zip_int64_t offset, int whence)
```
because the latest libzip sources appear to be not having  the old function
```C++
zip_fseek.c:58
ZIP_EXTERN int zip_fseek(struct zip_file *zf, long pos, int whence)
```
and there is no zip_fseek.c file at all in the latest libzip source directory (at least I didn't find that in my latest sources)
So the idea was to adapt the zip_source_seek function for calling from  byte_stream:844 line instead the old one. 
Also I'm taking into account your point about not to duplicate bytes_left variable if possible.
Please let me know if I missed something here...Thanks

**Comment by nleme** [comment]
@DennisPopovDn - unfortunately, it turned out to be too good to be true. I had started my tests with uncompressed content, since that is where the issue that triggered this whole investigation started (just see the title of this issue). However, once I moved forward to test content that was compressed, I found out that my solution would not work. :-( Let me explain why.

Let me refresh this for the people who were not at the last Readium SDK meeting. The current version of libzip is based upon the concept of "layered sources". You can interpret that as being a way to implement the "chain of responsibility" design pattern in pure C, without object orientation support. In it, each "layer" would behave like an object in the chain of responsibility, and each "layer" has a function associated to it that does the work that corresponds to each layer's purpose.

So, you start first with a raw file layer, which reads the raw file. Then, on top of that layer you add the "window" layer, which just keeps track of which ZIP entry in the ZIP file is currently being dealt with. This layer supports Seek and Tell, and things stop here when your content is not compressed. That corresponds to what I was seeing, and why this was working so fine.

However, things get messed up once you deal with compressed content. When that happens, libzip adds another layer on top of the "window" layer": it is the "deflate" layer. That layer will deal with decompressing the content. You can see it in the file zip_source_deflate.c file. It will use the deflate_decompress() function to deal with all the possible operations on the file. But, it explicitly rejects the Seek and Tell operations.

The end result is that, if you try to perform a Seek or Tell operation, it will simply return an error code.

Because of that, you cannot call zip_source_seek() on content that has been compressed, because it will just return an error code.

Currently, I'm stuck. The creators of libzip have gone out of their way to not support Seek and Tell. I already wrote to the libzip distribution list to see what is their take on this (I'll put here their answer as soon as I get it).

I read that you were trying to modify the code in the ZipFileByteStream to use zip_source_seek(). Based on what I saw in my own tests, that may not work on compressed content (if you find out differently, please let me know!). Do you have any other idea about how we can make Seek work with the latest version of libzip? Thanks for any help on this, and let me know if you need any help working on this.

**Comment by snuffin** [comment]
@nleme, the libzip discussion forum is still down and NIH hasn’t responded to our support request to fix it.  Maybe someone can suggest another way to get in touch with these guys?  In the meantime @DennisPopovDn can work on a prototype where we’ll modify the code in ZipFileByteStream to use zip_source_seek() per his comment above.  The prototype needs to support compressed and uncompressed content.  Let me know if that doesn’t make sense, otherwise we’ll start working on it this week.

**Comment by drminside** [comment]
DRM inside had the same problem during implementing the LCP on early this year. 

The problem was in that the fseek_bytes function (in the zip_fseek.c) could not calculate the zf->fpos and zf_cbytes_left values properly for the uncompressed resource.

We have fixed this matter with simply modifying inside of  the _zip_fseek_bytes() function.

**Comment by thkim2015** [comment]
Modified code is following;

/* seeking by raw byte amounts - no compression/decompression to handle */
int _zip_fseek_bytes(struct zip_file* zf, off_t abspos, off_t flen)
{
    /* can't set a negative offset */
    if (abspos < 0) {
        _zip_error_set(&zf->error, ZIP_ER_INVAL, 0);
        return -1;
    }
    /* CAN set offset past EOF: keeps offset, sets EOF */
    else if (abspos >= flen) {
        zf->flags |= ZIP_ZF_EOF;
        zf->bytes_left = 0;
        
        /* added by DRM inside, C.H. Yu on 2015-04-13 */
        // Without following added codes, uncompressed EPUB media content would not be properly random accessed,
        // such as bad index access error
        zf->fpos = _zip_file_get_offset_safe(zf->za, zf->file_index);
        zf->fpos += flen;
        zf->cbytes_left = 0;
        /* adding end */
    }
    /* not at or past EOF? ensure EOF is unset and update bytes_left */
    else {
        zf->flags &= ~ZIP_ZF_EOF;
        zf->bytes_left = flen - abspos;
        
        /* added by DRM inside, C.H. Yu on 2015-04-13 */
        // Without following added codes, uncompressed EPUB media content would not be properly random accessed,
        // such as bad index access error
        zf->fpos += (abspos - zf->file_fpos);
        zf->cbytes_left = zf->bytes_left;
        /* adding end */
    }
    zf->file_fpos = abspos;
    return 0;
}

**Comment by rkwright** [comment]
How does this compare to the solution in PR 179?  I think we should merge 179 if everyone agrees it is the right solution. I believe Bluefire has tested in thoroughly? This merge would resolve the issue but won't address the issue for files larger than 4 GB, but would be a good short-term solution, IMO.

**Comment by nleme** [comment]
Based on our latest discussion about this issue, currently I have been investigating how to bring the Seek functionality back into the latest version of libzip.

Just today we had a small piece of good news. Thomas Klausner, one of the authors of libzip, actually wrote back on a inquiry I had made some time ago. Originally, I had written about how to disable the CRC layer, given that that layer was returning that Seek was an unsupported operation. He answered back saying that we could pass a flag to disable that layer.

However, as I found out later, that is not the only obstacle. The deflate layer, the one that deals with compressed content, also blocks the Seek operation. Thomas made a small remark about that in this answer. Quote:

Currently, you can only seek in stored data (i.e., uncompressed data).
The deflate level does not support seeking.

I replied back to him thanking for answering our previous question, and then asked to him exactly about that. I.e., why compressed content does not support seeking and how hard it would be to bring it back. As soon as he answers back, I'll put his answer in this thread.

**Comment by nleme** [comment]
Good news: Thomas Klausner, from libzip, just answered back. He was quite nice. He confirmed what we suspected: Jim Dovey wrote the code for Seek in the old version of libzip.

Bad news: he stated that Seek was never supposed to be part of libzip in the first place, it was just added by Jim Dovey. As such, he said that they do not have plans to add that functionality back to libzip, and suggested that I ask Jim to add that functionality back to the current version of libzip.

**Comment by nleme** [comment]
What I meant above is that Thomas stated that it will be up to us to implement Seek in libzip for compressed content.

**Comment by nleme** [comment]
Another bit of information. I received this morning an e-mail from Dieter Baron, the other developer who works at libzip. Let me paste the important part of his e-mail below:

> We will try to add the Seek and Tell functionalities into the deflate layer.

Please don’t. Rather, write a seeking layer and put that on top if the underlying layer doesn’t support seeking directly.

However, be aware that seeking in a compressed stream is extremely inefficient: you can only read it from the beginning. So to seek back, you have to start over from the beginning and discard everything until you get to where you want.

If the files aren’t too big, it is much better to read them into a buffer and do random access on that.

(the phrase with the ">" was part of my original mail to the libzip authors)

In a nutshell, Dieter is saying to *not* add the Seek functionality to the deflate layer, but instead create a brand new layer (following their implementation of "chain of responsibility" in C) which would do the Seek.

He also goes back to a point that I've seen them doing before, i.e., suggesting that we do not do Seek at all, and instead just decompress the whole thing in a buffer.

Simply decompressing the whole file in a buffer is not something that is recommendable in mobile platforms: those files could be bigger than the available memory, especially in Android devices. Then, there is the possibility of writing the decompressed file to disk. The problem, then, is that we may be dealing with encrypted files, and you would not want to save in disk a decrypted file. Therefore, to be able to do this, we would have to do the following:

Decrypt ZIP entry -> Decompress ZIP entry -> Encrypt ZIP entry again -> Write ZIP entry to disk

Followed by:

Read ZIP entry from disk -> Decrypt ZIP entry -> Serve bits to Cocoa HTTP Server

That would be quite a change at this point.

**Comment by rkwright** [comment]
@nleme I agree with your assessment.  It sounds like the LibZip guys don't really understand the workflows we need to support.

**Comment by nleme** [comment]
@DennisPopovDn - how is your work on putting Seek back into libzip going? And how is that looking with regards to what Dieter said in his e-mail? I'm not demanding anything, especially because I'm aware that this is not an easy job, I'm just curious to know how the stuff you are doing is going to match with what Dieter said.

**Comment by danielweck** [comment]
> be aware that seeking in a compressed stream is extremely inefficient: you can only read it from the beginning. So to seek back, you have to start over from the beginning and discard everything until you get to where you want.

Correct. That's how I implemented byte stream "seek-ability" in SDKLauncher-OSX / ObjectiveC (a long time ago, before this feature was available natively in libzip), and that's how Jim Dovey implemented it as well (see `_zip_fseek_to_start` + `_zip_fseek_by_reading`):
https://github.com/readium/readium-sdk/blob/master/ePub3/ThirdParty/libzip/zip_fseek.c#L124

**Comment by danielweck** [comment]
Regarding the performance implication of seeking from the start of the byte stream: with both my old ObjectiveC implementation and Jim's current C implementation, seeking from the current position is an optimization that works in most streaming scenarios (e.g. video playing within an ebook page from start to end, or sequential audio Media Overlays playback), see:

https://github.com/readium/readium-sdk/blob/master/ePub3/ThirdParty/libzip/zip_fseek.c#L133

(note that `_zip_fseek_to_start` is not invoked when the cursor is positioned ahead)

```cpp
else if (abspos > zf->file_fpos) {
        // read & decompress bytes until we reach the right position
        return _zip_fseek_by_reading(zf, abspos-zf->file_fpos);
    }
```

**Comment by nleme** [comment]
To further comment on what Daniel said. We can put back the code for doing a Seek into libzip. There is the question, though, of how to do it.

If we modify the libzip code, then we can add a new layer as Dieter suggested. This way, it minimizes the changes on the existing files of libzip, which it would make easier to merge changes from libzip into our own version of libzip.

On the other hand, adding Seek into the deflate layer itself would be a larger change, and more difficult to merge with later changes into libzip.

Then, we can also add Seek as a wrapper around libzip. @danielweck, if I understood right, that is what you did way back in the day, right? Do you think that was a good approach?

**Comment by DennisPopovDn** [comment]
@nleme, I've been meditating on the libzip (how to add it into deflate layer with caching the data, w/o caching the data etc), and agreed with Dieter that implementing it as the separated layer seem to be the better idea... in progress

BTW I'm remembering some discussion at Readium about LCP decryption and byterange request (like a Seek feature), and it had been resolved with adding a cache buffer right to the some readiumsdk class.. it seems to me the buffering resolution has already been made for encrypted files.. May be it makes sense to implement that work around for compressed files as well?

**Comment by danielweck** [comment]
@nleme:
> Then, we can also add Seek as a wrapper around libzip. @danielweck, if I understood right, that is what you did way back in the day, right? Do you think that was a good approach?

The performance loss in my OSX-ObjectiveC implementation was probably due to creating / deleting ByteStream instances, which were needed in order to reset the zip file position to zero (equivalent to `_zip_fseek_to_start ` in the zip lib logic, see https://github.com/readium/readium-sdk/issues/177#issuecomment-120085546 ) when HTTP-requesting a byte range starting before the current stream cursor. Perhaps also because of the maintenance of intermediary byte buffers (I cannot remember exactly, but I'm pretty sure some "middle man" buffering mechanism was indeed needed).

Anyway, I guess here the proposal would be to implement the seek/tell functionality as "close" as possible to the native libzip code (very thin wrapper, no overhead, and crucially: no need for JNI byte buffer conversions in Android).

**Comment by nleme** [comment]
@DennisPopovDn - you are correct: we fixed one issue that we had around *non-byte-range* requests in the Readium SDK by making a cache. That fixed some big problems that we had. However, I don't think we can apply the same solution to the byte-range request case because of the mobile platforms. In that case, the problem is that you may have a really big media file that, if loaded in a buffer in memory, could possibly completely fill out the memory of the device (or not even fit) especially in some of the less powerful Android devices.

@danielweck - I see what you are saying. Very thin wrapper *outside* of the libzip code, just to keep track of the reading position (Tell) and to perform the required read operations (Seek). That is something that we can try.

**Comment by danielweck** [comment]
Is this issue now fixed? See already-merged PR from DRM-Inside:
https://github.com/readium/readium-sdk/pull/219/files#diff-e8c489d80c21c04a0f36514e70875fd3

**Comment by drminside** [comment]
@danielweck - With merged codes from DRM inside's PR, I think this issue is fixed.
Because the reason why DRM inside modified the zip_fseek.c was that we had the same problem with what  @nleme mentioned in the issue #177.

We had verified the validity of the modified code with Readium Launcher and several uncompressed media resources in the zip file.

**Comment by danielweck** [comment]
I think this issue can be closed, as well as @nleme 's associated PR (superseded by @drminside 's merged PR).
See:

https://github.com/readium/readium-sdk/pull/179

https://github.com/readium/readium-sdk/pull/219/files#diff-e8c489d80c21c04a0f36514e70875fd3
